metrics/influx/influx.go

```diff
@@ -0,0 +1,213 @@
+}
+
+type InfluxMetricsEntryByFormat struct {
+	Duration       metrics.Timer
+	Request        metrics.Meter
+	Errors         metrics.Meter
+	BadRequest     metrics.Meter
+	JsonRequest    metrics.Meter
+	XmlRequest     metrics.Meter
+	DefinesTTL     metrics.Meter
+	InvalidRequest metrics.Meter
+	RequestLength  metrics.Histogram
+}
+
+type InfluxConnectionMetrics struct {
+	ActiveConnections      metrics.Counter
+	ConnectionCloseErrors  metrics.Meter
+	ConnectionAcceptErrors metrics.Meter
+}
+type InfluxExtraTTL struct {
+	ExtraTTLSeconds metrics.Histogram
+}
+
+func NewInfluxMetricsEntry(name string, r metrics.Registry) *InfluxMetricsEntry {
+	return &InfluxMetricsEntry{
+		Duration:   metrics.GetOrRegisterTimer(fmt.Sprintf("%s.request_duration", name), r),
+		Errors:     metrics.GetOrRegisterMeter(fmt.Sprintf("%s.error_count", name), r),
+		BadRequest: metrics.GetOrRegisterMeter(fmt.Sprintf("%s.bad_request_count", name), r),
+		Request:    metrics.GetOrRegisterMeter(fmt.Sprintf("%s.request_count", name), r),
+	}
+}
+
+func NewInfluxMetricsEntryBackendPuts(name string, r metrics.Registry) *InfluxMetricsEntryByFormat {
+	return &InfluxMetricsEntryByFormat{
+		Duration:       metrics.GetOrRegisterTimer(fmt.Sprintf("%s.request_duration", name), r),
+		Errors:         metrics.GetOrRegisterMeter(fmt.Sprintf("%s.error_count", name), r),
+		BadRequest:     metrics.GetOrRegisterMeter(fmt.Sprintf("%s.bad_request_count", name), r),
+		JsonRequest:    metrics.GetOrRegisterMeter(fmt.Sprintf("%s.json_request_count", name), r),
+		XmlRequest:     metrics.GetOrRegisterMeter(fmt.Sprintf("%s.xml_request_count", name), r),
+		DefinesTTL:     metrics.GetOrRegisterMeter(fmt.Sprintf("%s.defines_ttl", name), r),
+		InvalidRequest: metrics.GetOrRegisterMeter(fmt.Sprintf("%s.unknown_request_count", name), r),
+		RequestLength:  metrics.GetOrRegisterHistogram(name+".request_size_bytes", r, metrics.NewExpDecaySample(1028, 0.015)),
+	}
+}
+
+func NewInfluxConnectionMetrics(r metrics.Registry) *InfluxConnectionMetrics {
+	return &InfluxConnectionMetrics{
+		ActiveConnections:      metrics.GetOrRegisterCounter("connections.active_incoming", r),
+		ConnectionAcceptErrors: metrics.GetOrRegisterMeter("connections.accept_errors", r),
+		ConnectionCloseErrors:  metrics.GetOrRegisterMeter("connections.close_errors", r),
+	}
+}
+
+func CreateInfluxMetrics() *InfluxMetrics {
+	flushTime := TenSeconds
+	r := metrics.NewPrefixedRegistry("prebidcache.")
+	m := &InfluxMetrics{
+		Registry:    r,
+		Puts:        NewInfluxMetricsEntry("puts.current_url", r),
+		Gets:        NewInfluxMetricsEntry("gets.current_url", r),
+		PutsBackend: NewInfluxMetricsEntryBackendPuts("puts.backend", r),
+		GetsBackend: NewInfluxMetricsEntry("gets.backend", r),
+		Connections: NewInfluxConnectionMetrics(r),
+		ExtraTTL:    &InfluxExtraTTL{ExtraTTLSeconds: metrics.GetOrRegisterHistogram("extra_ttl_seconds", r, metrics.NewUniformSample(5000))},
+	}
+
+	metrics.RegisterDebugGCStats(m.Registry)
```

Comment:
Prometheus is currently missing these. Please check if you can get these all from using `NewGoCollector()`.
https://github.com/prometheus/client_golang/blob/aa9238db679fc02bf22cf4f2c27a980edcb5ada0/prometheus/go_collector.go#L64

Reply:
I was about to add both `prometheus.NewProcessCollector` and `prometheus.NewGoCollector()` because to the best of my knowledge they are the equivalent of `metrics.CaptureRuntimeMemStats` and `metrics.CaptureDebugGCStats` on the influx side. But at the end I didn't add them because, according to GoDoc site, these two already come in the default registry as described in https://godoc.org/github.com/prometheus/client_golang/prometheus (paragraph shown below). Therefore, I believe CPU usage, and Ram can be collected without the need of explicitly declaring them nor assigning them to the custom `registry` we created in line 63 of `metrics/prometheus/prometheus.go`. Further evidence of this hypothesis is the fact that in `prebid-server` side neither `prometheus.NewProcessCollector` nor `prometheus.NewGoCollector` are ever found in the code base and still we effectively collect CPU and Ram usage for every pod.
> "Also note that the DefaultRegisterer comes registered with a Collector for Go runtime metrics (via NewGoCollector) and a Collector for process metrics (via NewProcessCollector). With a custom registry, you are in control and decide yourself about the Collectors to register."

At the end I didn't add these two at the bottom of `CreatePrometheusMetrics` found in `metrics/prometheus/prometheus.go`. Please let me know if you are ok with this or if you want me to explicitly add them.
```
	// Should be the equivalent of the following influx collectors
	// go metrics.CaptureRuntimeMemStats(m.Registry, flushTime)
	// go metrics.CaptureDebugGCStats(m.Registry, flushTime)
	registry.MustRegister(
		prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{Namespace: cfg.Namespace}),
		prometheus.NewGoCollector(),
	)
```

Reply:
We need to explicitly register the `NewProcessCollector` and the `NewGoCollector` since we're using a custom registry instead of the default registry.

Reply:
Done